src/rpc/bootstrap_daemon.h

```diff
@@ -0,0 +1,184 @@
+    }
+
+    template <class t_request, class t_response>
+    bool invoke_http_json_rpc(const boost::string_ref command_name, const t_request &out_struct, t_response &result_struct) noexcept
+    {
+      if (!switch_daemon_if_needed())
+      {
+        return false;
+      }
+
+      epee::json_rpc::request<t_request> json_req = AUTO_VAL_INIT(json_req);
+      epee::json_rpc::response<t_response, std::string> json_resp = AUTO_VAL_INIT(json_resp);
+      json_req.jsonrpc = "2.0";
+      json_req.id = epee::serialization::storage_entry(0);
+      json_req.method = std::string(command_name.begin(), command_name.end());
+      json_req.params = out_struct;
+
+      if (!handle_result(epee::net_utils::invoke_http_json("/json_rpc", json_req, json_resp, m_http_client)))
+      {
+        return false;
+      }
+
+      result_struct = json_resp.result;
+      return true;
+    }
+
+  private:
+    bool set_server(const std::string &address, const boost::optional<epee::net_utils::http::login> &credentials) noexcept
+    {
+      if (!m_http_client.set_server(address, credentials))
+      {
+        MERROR("Failed to set bootstrap daemon address " << address);
+        return false;
+      }
+
+      m_address = address;
+      MINFO("Changed bootstrap daemon address to " << m_address);
+
+      return true;
+    }
+
+    const nodetool::peerlist_entry *get_random_public_node(const std::vector<nodetool::peerlist_entry> &peers) const noexcept
+    {
+      const size_t size = peers.size();
+      size_t cursor = crypto::rand_idx(size);
+      for (size_t index = 0; index < size; ++index)
+      {
+        const auto *peer = &peers[++cursor % size];
+        if (peer->rpc_port != 0)
+        {
+          return peer;
+        }
+      }
+
+      return nullptr;
+    }
+
+    bool switch_daemon_if_needed() noexcept
+    {
+      if (m_p2p == nullptr || m_last_result)
+      {
+        return true;
+      }
+
+      std::vector<nodetool::peerlist_entry> white_list;
+      std::vector<nodetool::peerlist_entry> gray_list;
+      m_p2p->get_public_peerlist(gray_list, white_list);
+
+      const nodetool::peerlist_entry *public_node = get_random_public_node(white_list);
+      if (public_node == nullptr)
+      {
+        MDEBUG("No white public node found, checking gray peers");
+
+        public_node = get_random_public_node(gray_list);
+        if (public_node == nullptr)
+        {
+          MERROR("Failed to switch bootstrap daemon, no white or gray public node found");
+          return false;
+        }
+      }
+
+      return set_server(public_node->adr.host_str() + ":" + std::to_string(public_node->rpc_port), boost::none);
+    }
+
+  private:
+    std::string m_address;
+    epee::net_utils::http::http_simple_client m_http_client;
+    bool m_last_result;
+    nodetool::node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core>> *m_p2p;
```

Comment:
The lifetime/ownership of this is unfortunate (but acceptable for its only current use case), no unit tests were added, and there is little code (outside of the randomized peer selection which can still be in a separate file) ... should the bulk of this logic just be included in the `core_rpc_server.cpp`?